tests/func/test_add.py

```diff
@@ -861,7 +861,7 @@ def test_add_optimization_for_hardlink_on_empty_files(tmp_dir, dvc, mocker):
     m = mocker.spy(LocalFileSystem, "is_hardlink")
     stages = dvc.add(["foo", "bar", "lorem", "ipsum"])
 
-    assert m.call_count == 1
+    assert m.call_count == 4
```

Comment:
Previously we would cache link type for the whole cache, which was incorrect, so now we'll do that for every object - which is correct, but not super optimal for large number of artifacts (but optimal for big datasets). More optimal solution will require some kind of prefix caching, which is not worth getting into right now.